cryptofeed/exchange/okcoin.py

```diff
@@ -54,32 +56,60 @@ def __calc_checksum(self, pair):
-            if "SWAP" in pair:
-                return channel.format('swap')
-            elif len(pair.split("-")) == 3:
-                return channel.format('futures')
-            else:
-                return channel.format('spot')
-
+    @staticmethod
+    def instrument_type(pair):
+        dash_count = pair.count("-")
+        if dash_count == 1: # BTC-USDT
+            return 'spot'
+        if dash_count == 4:  # BTC-USD-201225-35000-P
+            return 'option'
+        if pair[-4:] == "SWAP":  # BTC-USDT-SWAP
+            return 'swap'
+        return 'futures'  # BTC-USDT-201225
+
+    def get_channel_pairs(self):
         if self.config:
             for chan in self.config:
                 if chan == LIQUIDATIONS:
                     continue
-                args = [f"{chan_format(chan, pair)}:{pair}" for pair in self.config[chan]]
-                await websocket.send(json.dumps({
-                    "op": "subscribe",
-                    "args": args
-                }))
+                for pair in self.config[chan]:
+                    yield chan, pair
         else:
-            chans = [f"{chan_format(chan, pair)}:{pair}" for chan in self.channels for pair in self.pairs if chan != LIQUIDATIONS]
-            await websocket.send(json.dumps({
-                "op": "subscribe",
-                "args": chans
-            }))
+            for chan in self.channels:
+                if chan == LIQUIDATIONS:
+                    continue
+                for pair in self.pairs:
+                    yield chan, pair
+
+    def get_channel_pair_combinations(self):
+        for chan, pair in self.get_channel_pairs():
+            instrument_type = self.instrument_type(pair)
+            if instrument_type != 'swap' and 'funding' in chan:
+                continue  # No funding for spot, futures and options
+            yield f"{chan.format(instrument_type)}:{pair}"
+
+    async def subscribe(self, websocket: websockets.WebSocketClientProtocol) -> bool:
+        self.__reset()
+
+        args_list = list(self.get_channel_pair_combinations())
+        LOG.info("%s: Got %r combinations of pairs and channels", self.id, len(args_list))
+
+        if len(args_list) == 0:
+            LOG.info("%s: No websocket subscription", self.id)
+            return False
+
+        # Avoid error "Max frame length of 65536 has been exceeded" by limiting requests to some args
+        smaller_lists = split.limit_list_size(args_list, 33)
+
+        for args in smaller_lists:
+            LOG.info("%s: Subscribe to %s args from %r to %r", self.id, len(args), args[0], args[-1])
+            request = {"op": "subscribe", "args": args}
+            await websocket.send(json.dumps(request))
+            # wait for response (max 1 second) before next subscription batch
+            msg = await asyncio.wait_for(websocket.recv(), timeout=1)
+            if msg:
+                await self.message_handler(msg, time.time_ns())
+        return True
```

Comment:
this code in subscribe seems interesting and useful. I'm not sure of the problem its solving, but would be nice to get this in.w

Reply:
This code avoids flooding exchange with too many subscribe requests: we wait at least one response with one second timeout. And this code avoids buffering received responses.

Reply:
looks pretty good, do you want to port this change over individually? I'm happy to copy/paste it and then test it and get it merged, but its your code and contribution so I'm fine waiting for you to do it if you want to.

Reply:
:ok: :+1: I am preparing a PR to propose a refactoring/cleaning your new design using `AsyncConnection`

Reply:
:+1:

Reply:
Happy New Year Bryant @bmoscon 

I have worked hard to improve your new `AsyncConnection` design and I will propose you a nice PR with a new file `runner.py` implementing the class `Runner`. The `Runner` owns the main `message_handler` infinite loop ( and `_watch()`) that was previously the responsibility of `FeedHandler`.

In the coming PR, I have replaced the `Feed.connections()` function by reusing the `address: dict` in the `Feed.__init__()` (first argument). No need to pass `partial(subscribe,x)`
 and `message_handler()` callbacks. The `address: dict` (options and URLs) are used by the `FeedHandler` to create the `AsyncConnection`: each `AsyncConnection` keeps its option/address within its context. This connection context is then accessed by the `Feed` subclass to subscribe to the right pairs/channels.

The `FeedHandler` manages the following objects:
* `Feed` (exchange or provider)
* `AsyncConnection` (Websocket or HTTP)
* `Runner` (one `Runner` per `AsyncConnection`)

The `FeedHandler` has the responsibility to create the `Feed` subsclass, to create its the corresponding `AsyncConnection` according to the `Feed.address` (`str` or `dict`, can be a mix of WS and HTTP), and to create one `Runner` per `AsyncConnection`. Each `Feed` has one or more `AsyncConnection`. Each `Runner` runs only one `AsyncConnection`.

`Runner` knows/calls `Feed` and `AsyncConnection`. But the `Feed` and its subclasses do not know the `Runner`. The `Runner` pass the `AsyncConnection` when calling `Feed.subscribe() and `Feed.handle()`. The `Feed` subclasses use the `AsyncConnection` to access the connection context and to send messages (websocket only). Notice I have renamed `message_handler()` -> `handle()` for consistency with `subscribe() != message_subscriber()`.

Once everything is created, the `FeedHandler` runs all the `Runner` asynchronously. The `Runner` infinite loop calls the `Feed.subscribe(conn)`, and loops over the `AsyncConnection.read()`. Each received message is passed to `Feed.handle(msg,timestamp,conn)`. The `Feed` subcalss has access to `conn.ctx`, the connection context: no need to mange a `seqno_map[conn.id]`. The `conn.ctx` contains the `Feed.address` (and options) used to create the `AsyncConnection`.

All exchanges using websocket benefits this feature:
```py
# slow down the subscription batch by waiting for a response or for a half of a second before next send
try:
    msg = await asyncio.wait_for(self.socket.recv(), timeout=0.5)
    await self.handle(msg, time(), self)
except asyncio.exceptions.TimeoutError:
    pass
```

On shutdown signal, the `FeedHandler` closes the `AsyncConnection`, stops the `Runner` & `Feed`, and waits for termination of all backends before exiting. The `FeedHandler` runs this functions using AsyncIO to speed up the shutdown process. This is important because multiple backends may need to flush their data and the orchestrator (K8s/Docker) may kill Cryptofeed is it takes too much time to terminate. However, even if the backend flushing is faster, `loop.run_until_complete(asyncio.gather(*flush_tasks))` takes some seconds to detect all flushing tasks are completed.

I have adapted almost all exchanges and providers: I have to finish WHALES_ALERT. This improvement of your new design simplifies the source code, especially BITFINEX, BINANCE and more... I have also tested long run with all exchanges/providers activated (except WHALES_ALERT), subscribing to many many pairs per `Feed`, and with backends requiring final flush.

PR will be ready at the end of the week...

Reply:
@olibre  bonne année to you as well. Looking forward to it!